From a76ce150fb78811368068e02d5f5050a551541a9 Mon Sep 17 00:00:00 2001
From: Fei Wang <fei.w.wang@intel.com>
Date: Fri, 17 Apr 2020 10:46:35 +0800
Subject: [PATCH 1/3] lavu/pix_fmt: add new pixel format x2rgb10

The format is packed RGB with each channel 10 bits available and
include 2 bits unused.

Signed-off-by: Fei Wang <fei.w.wang@intel.com>
---
 libavutil/pixdesc.c                     | 24 ++++++++++++++++++++++++
 libavutil/pixfmt.h                      |  3 +++
 tests/ref/fate/filter-pixdesc-x2rgb10le |  1 +
 tests/ref/fate/sws-pixdesc-query        | 11 +++++++++++
 4 files changed, 39 insertions(+)
 create mode 100644 tests/ref/fate/filter-pixdesc-x2rgb10le

diff --git a/libavutil/pixdesc.c b/libavutil/pixdesc.c
index 89075e60f4..e7d42654b4 100644
--- a/libavutil/pixdesc.c
+++ b/libavutil/pixdesc.c
@@ -289,6 +289,30 @@ static const AVPixFmtDescriptor av_pix_fmt_descriptors[AV_PIX_FMT_NB] = {
         },
         .flags = AV_PIX_FMT_FLAG_RGB,
     },
+    [AV_PIX_FMT_X2RGB10LE] = {
+        .name = "x2rgb10le",
+        .nb_components= 3,
+        .log2_chroma_w= 0,
+        .log2_chroma_h= 0,
+        .comp = {
+            { 0, 4, 2, 4, 10, 3, 9, 2 }, 	   /* R */
+            { 0, 4, 1, 2, 10, 3, 9, 3 }, 	   /* G */
+            { 0, 4, 0, 0, 10, 3, 9, 4 }, 	   /* B */
+        },
+        .flags = AV_PIX_FMT_FLAG_RGB,
+    },
+    [AV_PIX_FMT_X2RGB10BE] = {
+        .name = "x2rgb10be",
+        .nb_components= 3,
+        .log2_chroma_w= 0,
+        .log2_chroma_h= 0,
+        .comp = {
+            { 0, 4, 0, 4, 10, 3, 9, 2 }, 	   /* R */
+            { 0, 4, 1, 2, 10, 3, 9, 3 }, 	   /* G */
+            { 0, 4, 2, 0, 10, 3, 9, 4 }, 	   /* B */
+        },
+        .flags = AV_PIX_FMT_FLAG_RGB | AV_PIX_FMT_FLAG_BE,
+    },
     [AV_PIX_FMT_YUV422P] = {
         .name = "yuv422p",
         .nb_components = 3,
diff --git a/libavutil/pixfmt.h b/libavutil/pixfmt.h
index 4264c45abc..1af1ec7cd5 100644
--- a/libavutil/pixfmt.h
+++ b/libavutil/pixfmt.h
@@ -362,6 +362,8 @@ enum AVPixelFormat {
     AV_PIX_FMT_Y410LE,    ///< packed YUV 4:4:4, 32bpp, Cr  Y Cb  A, little-endian
     AV_PIX_FMT_Y410BE,    ///< packed YUV 4:4:4, 32bpp, Cr  Y Cb  A, big-endian
 
+    AV_PIX_FMT_X2RGB10LE,    ///< packed RGB 10:10:10, 30bpp, (msb)2X 10R 10G 10B(lsb), little-endian, X=unused/undefined
+    AV_PIX_FMT_X2RGB10BE,    ///< packed RGB 10:10:10, 30bpp, (msb)2X 10R 10G 10B(lsb), big-endian, X=unused/undefined
     AV_PIX_FMT_NB         ///< number of pixel formats, DO NOT USE THIS if you want to link with shared libav* because the number of formats might differ between versions
 };
 
@@ -452,6 +454,7 @@ enum AVPixelFormat {
 
 #define AV_PIX_FMT_Y210       AV_PIX_FMT_NE(Y210BE,  Y210LE)
 #define AV_PIX_FMT_Y410       AV_PIX_FMT_NE(Y410BE,  Y410LE)
+#define AV_PIX_FMT_X2RGB10       AV_PIX_FMT_NE(X2RGB10BE, X2RGB10LE)
 
 /**
   * Chromaticity coordinates of the source primaries.
diff --git a/tests/ref/fate/filter-pixdesc-x2rgb10le b/tests/ref/fate/filter-pixdesc-x2rgb10le
new file mode 100644
index 0000000000..94c8640a56
--- /dev/null
+++ b/tests/ref/fate/filter-pixdesc-x2rgb10le
@@ -0,0 +1 @@
+pixdesc-x2rgb10le    98d697ed4668daf535163d5e08c903bb
diff --git a/tests/ref/fate/sws-pixdesc-query b/tests/ref/fate/sws-pixdesc-query
index d0efa92671..f023430356 100644
--- a/tests/ref/fate/sws-pixdesc-query
+++ b/tests/ref/fate/sws-pixdesc-query
@@ -57,6 +57,8 @@ isNBPS:
   nv20le
   p010be
   p010le
+  x2rgb10be
+  x2rgb10le
   xyz12be
   xyz12le
   y210be
@@ -143,6 +145,7 @@ isBE:
   rgb555be
   rgb565be
   rgba64be
+  x2rgb10be
   xyz12be
   y210be
   y410be
@@ -442,6 +445,8 @@ isRGB:
   rgb8
   rgba64be
   rgba64le
+  x2rgb10be
+  x2rgb10le
 
 Gray:
   gray
@@ -588,6 +593,8 @@ AnyRGB:
   rgb8
   rgba64be
   rgba64le
+  x2rgb10be
+  x2rgb10le
 
 ALPHA:
   ayuv64be
@@ -698,6 +705,8 @@ Packed:
   rgba64le
   uyvy422
   uyyvyy411
+  x2rgb10be
+  x2rgb10le
   xyz12be
   xyz12le
   y210be
@@ -864,6 +873,8 @@ PackedRGB:
   rgb8
   rgba64be
   rgba64le
+  x2rgb10be
+  x2rgb10le
 
 PlanarRGB:
   gbrap
-- 
2.17.1

